google/cloud/spanner/internal/session_pool.cc

```diff
@@ -166,10 +165,8 @@ void SessionPool::Release(std::unique_ptr<Session> session) {
     --total_sessions_;
     return;
   }
-  bool notify = sessions_.empty();
   sessions_.push_back(std::move(session));
-  // If sessions_ was empty, wake up someone who was waiting for a session.
-  if (notify) {
+  if (num_waiting_for_session_ > 0) {
```

Comment:
I'm fine with this ... while it does not eliminate the possibility of useless (and purportedly expensive) notifications, it does reduce them.

And all the better if you can use `num_waiting_for_session_` to make `CreateSessions()` a little smarter.